src/broker/cas_cgw.c

```diff
@@ -2438,3 +2455,277 @@ cgw_utype_to_string (int type)
       return (char *) "";
     }
 }
+
+char *
+cgw_rewrite_query (char *src_query, char *cublink_pos)
+{
+  char *rewrite_query = NULL;
+  char *select_from = NULL;
+  char *inline_view = NULL;
+  char *new_inline_view = NULL;
+  char *cublink = NULL;
+  char *cols_type = NULL;
+  char **cols_list = NULL;
+  char *where = NULL;
+
+  char *start = NULL;
+  char *end = NULL;
+
+  size_t str_size = 0;
+  SQLSMALLINT num_cols = 0;
+  bool is_rewrite_error = false;
+  int err_code = 0;
+
+  int cublink_col_list_len = 0;
+  int where_len = 0;
+  int inline_view_len = 0;
+  int new_inline_view_len = 0;
+
+  char col_name[COL_NAME_LEN + 1];
+  SQLSMALLINT col_name_length = 0;
+  SQLHSTMT hstmt;
+
+  start = strstr (src_query, REWRITE_DELIMITER_FROM);
+  if (start == NULL)
+    {
+      is_rewrite_error = true;
+      goto REWRITE_ERROR;
+    }
+
+  str_size = ((start + REWRITE_DELIMITER_FROM_LEN) - src_query);
+  select_from = (char *) MALLOC (str_size + 1);
+  strncpy (select_from, src_query, str_size);
+  select_from[str_size] = '\0';
+
+  end = cublink_pos;
+  if (cublink_pos == NULL)
+    {
+      is_rewrite_error = true;
+      goto REWRITE_ERROR;
+    }
+
+  inline_view_len = (end + strlen (") ")) - (start + REWRITE_DELIMITER_FROM_LEN) + 1;
+  inline_view = (char *) MALLOC (inline_view_len);
+  strncpy (inline_view, start + REWRITE_DELIMITER_FROM_LEN, inline_view_len - 1);
+  inline_view[inline_view_len - 1] = '\0';
+
+  start = cublink_pos;
+  if (start == NULL)
+    {
+      is_rewrite_error = true;
+      goto REWRITE_ERROR;
+    }
+
+  end = strstr (start, REWRITE_DELIMITER_WHERE);
+  if (end == NULL)
+    {
+      is_rewrite_error = true;
+      goto REWRITE_ERROR;
+    }
+
+  str_size = end - (start + strlen (") "));
+  cublink = (char *) MALLOC (str_size + 1);
+  strncpy (cublink, (start + strlen (") ")), str_size);
+  cublink[str_size] = '\0';
+
+  cublink_col_list_len = end - (start + REWRITE_DELIMITER_CUBLINK_LEN) + 1;
+  cols_type = (char *) MALLOC (cublink_col_list_len);
+  strncpy (cols_type, (start + REWRITE_DELIMITER_CUBLINK_LEN), cublink_col_list_len - 1);
+  cols_type[cublink_col_list_len - 1] = '\0';
+
+  cgw_eliminate (cols_type, '(');
+  cgw_eliminate (cols_type, ')');
+  cols_list = cgw_split_string (cols_type, ",");
```

Comment:
the number of columns can be calculated as below:
`for (t = strtok(cols_type, ","), col_cnt =0; t = strtok(NULL, ","); con_cnt++)  ;`
after getting the count, it can be compared to the column count of original query from SQLPreare for error check